File: ios/ComapeoPrefs.swift
import Foundation

/// Persistent storage for sentry-related user preferences. Snapshot-
/// at-launch: toggle changes take effect on next launch; flipping a
/// toggle to `false` also wipes the sentry-cocoa cache so queued events
/// don't ship. Mirrors `ComapeoPrefs.kt`.
///
/// Persistence goes through a `Store` so unit tests can back it with a
/// plain dictionary instead of a real `UserDefaults`.
final class ComapeoPrefs {
    /// Minimal persistence surface. `nil` from a getter means "key
    /// absent" (so the caller falls back to a default). Production wraps
    /// `UserDefaults`; tests use an in-memory dictionary.
    protocol Store {
        func getBool(_ key: String) -> Bool?
        func setBool(_ key: String, _ value: Bool)
        func getDouble(_ key: String) -> Double?
        func setDouble(_ key: String, _ value: Double)
        func getString(_ key: String) -> String?
        func setString(_ key: String, _ value: String)
        func remove(_ key: String)
    }

    struct Defaults {
        let diagnosticsEnabled: Bool
        let applicationUsageData: Bool
        let debug: Bool
    }

    private let store: Store
    private let defaults: Defaults
    /// Wall clock; injectable so the debug-auto-off tests don't depend on real time.
    private let now: () -> Double

    init(
        store: Store,
        defaults: Defaults,
        now: @escaping () -> Double = { Date().timeIntervalSince1970 * 1000 }
    ) {
        self.store = store
        self.defaults = defaults
        self.now = now
    }

    func readDiagnosticsEnabled() -> Bool {
        store.getBool(Key.diagnosticsEnabled) ?? defaults.diagnosticsEnabled
    }

    func readApplicationUsageData() -> Bool {
        store.getBool(Key.applicationUsageData) ?? defaults.applicationUsageData
    }

    /// Raw stored `debug` value with no auto-off side effect — the user's saved
    /// toggle for a live settings read. `readDebugEnabled()` applies the 72h
    /// auto-off (and its disk mutation) at launch; this must not, so a getter
    /// never writes.
    func readDebugStored() -> Bool {
        store.getBool(Key.debug) ?? defaults.debug
    }

    /// Read the `debug` toggle, applying the `debugMaxAgeMs` auto-off: if
    /// debug was switched on longer ago than that, flip it off, clear the
    /// timestamp, queue a `comapeo.debug.auto_disabled` breadcrumb, and
    /// return `false`. A `debug=true` cell with no timestamp (e.g. enabled
    /// via the configured default) is treated as "enabled now" and stamped
    /// on first read.
    ///
    /// The window is wall-clock based (it must survive process restarts),
    /// so a backward clock change is treated conservatively: an enable
    /// timestamp in the future expires debug rather than extending it. This
    /// is a best-effort privacy window on the user's own device, not a
    /// security boundary.
    func readDebugEnabled() -> Bool {
        let stored = store.getBool(Key.debug) ?? defaults.debug
        if !stored { return false }
        guard let enabledAt = store.getDouble(Key.debugEnabledAtMs) else {
            store.setDouble(Key.debugEnabledAtMs, now())
            return true
        }
        let age = now() - enabledAt
        if age < 0 || age > Self.debugMaxAgeMs {
            store.setBool(Key.debug, false)
            store.remove(Key.debugEnabledAtMs)
            DebugAutoOff.queueBreadcrumb()
            return false
        }
        return true
    }

    func writeDiagnosticsEnabled(_ value: Bool) {
        store.setBool(Key.diagnosticsEnabled, value)
    }

    func writeApplicationUsageData(_ value: Bool) {
        store.setBool(Key.applicationUsageData, value)
    }

    /// The permanent per-install root user ID (a short `XXXX-XXXX-XXXX`
    /// code — see `SentryUserId.generateRootId`), generated lazily on first
    /// read. Never sent to Sentry — Sentry `user.id` values are derived from
    /// it via `SentryUserId.derive`. Exposed to the host app (via
    /// `getSentryRootUserId`) so a user can share it for debugging and we can
    /// recompute their historical monthly IDs. Lives in `UserDefaults` (not
    /// Keychain) deliberately: uninstall should genuinely reset identity.
    func readRootUserId() -> String {
        if let existing = store.getString(Key.rootUserId) { return existing }
        let generated = SentryUserId.generateRootId()
        store.setString(Key.rootUserId, generated)
        return generated
    }

    /// The Sentry `user.id` for this launch: permanent when the user opted
    /// in to application-usage data, otherwise rotating monthly (UTC).
    func deriveSentryUserId(applicationUsageData: Bool) -> String {
        SentryUserId.derive(
            rootUserId: readRootUserId(),
            permanent: applicationUsageData,
            nowMs: now()
        )
    }

    /// Write `debug`, stamping (true) or clearing (false) the enable
    /// timestamp synchronously. Re-writing `true` refreshes the window.
    func writeDebugEnabled(_ value: Bool) {
        store.setBool(Key.debug, value)
        if value {
            store.setDouble(Key.debugEnabledAtMs, now())
        } else {
            store.remove(Key.debugEnabledAtMs)
        }
    }

    enum Key {
        static let diagnosticsEnabled = "sentry.diagnosticsEnabled"
        static let applicationUsageData = "sentry.applicationUsageData"
        static let debug = "sentry.debug"
        static let debugEnabledAtMs = "sentry.debugEnabledAtMs"
        static let rootUserId = "sentry.rootUserId"
    }

    /// 72h in milliseconds — debug mode auto-disables this long after enable.
    static let debugMaxAgeMs: Double = 72 * 60 * 60 * 1000

    /// Privacy model treats baseline error visibility as on.
    static let defaultDiagnosticsEnabled: Bool = true
    /// Off until user opts in.
    static let defaultApplicationUsageData: Bool = false
    static let defaultDebug: Bool = false

    static func open() -> ComapeoPrefs {
        let sentryConfig = SentryConfig.loadFromMainBundle()
        let defaults = Defaults(
            diagnosticsEnabled: sentryConfig?.diagnosticsEnabledDefault
                ?? defaultDiagnosticsEnabled,
            applicationUsageData: sentryConfig?.applicationUsageDataDefault
                ?? defaultApplicationUsageData,
            debug: sentryConfig?.debugDefault ?? defaultDebug
        )
        return ComapeoPrefs(
            store: UserDefaultsStore(defaults: UserDefaults.standard),
            defaults: defaults
        )
    }

    /// `UserDefaults`-backed [Store]. `object(forKey:)` distinguishes absent
    /// from an explicit `false`; `bool(forKey:)` collapses them, which would
    /// silently re-enable diagnostics on every user `false`.
    private struct UserDefaultsStore: Store {
        let defaults: UserDefaults
        func getBool(_ key: String) -> Bool? { defaults.object(forKey: key) as? Bool }
        func setBool(_ key: String, _ value: Bool) { defaults.set(value, forKey: key) }
        func getDouble(_ key: String) -> Double? { defaults.object(forKey: key) as? Double }
        func setDouble(_ key: String, _ value: Double) { defaults.set(value, forKey: key) }
        func getString(_ key: String) -> String? { defaults.string(forKey: key) }
        func setString(_ key: String, _ value: String) { defaults.set(value, forKey: key) }
        func remove(_ key: String) { defaults.removeObject(forKey: key) }
    }

    /// Recursively delete sentry-cocoa's on-disk cache root at
    /// `<NSCachesDirectory>/io.sentry/` (sentry-cocoa's documented
    /// default). Wipes envelopes, sessions, and scope state so a
    /// `diagnosticsEnabled=false` flip can't ship anything from the
    /// current session.
    ///
    /// Best-effort: filesystem errors don't block the opt-out. Worst
    /// case the cache survives one more launch, where diagnostics is
    /// off so nothing reads it.
    static func wipeSentryOutbox() {
        guard let caches = FileManager.default
            .urls(for: .cachesDirectory, in: .userDomainMask)
            .first else { return }
        wipeSentryOutbox(at: caches.appendingPathComponent("io.sentry", isDirectory: true))
    }

    /// Path-taking variant for unit tests. Missing directory is success.
    static func wipeSentryOutbox(at url: URL) {
        try? FileManager.default.removeItem(at: url)
    }
}

/// Holds the `comapeo.debug.auto_disabled` breadcrumb queued by the debug
/// auto-off. `readDebugEnabled()` runs before `SentrySDK.start`,
/// so the breadcrumb can't be added directly; it's drained once the SDK
/// is up.
enum DebugAutoOff {
    private static let lock = NSLock()
    private static var pendingFlag = false

    static func queueBreadcrumb() {
        lock.lock()
        pendingFlag = true
        lock.unlock()
    }

    /// Consume the pending flag; returns whether a breadcrumb is owed.
    static func consume() -> Bool {
        lock.lock()
        defer { lock.unlock() }
        let owed = pendingFlag
        pendingFlag = false
        return owed
    }
}

File: android/src/main/java/com/comapeo/core/ComapeoPrefs.kt
package com.comapeo.core

import android.content.Context
import android.content.SharedPreferences
import androidx.core.content.edit
import java.io.File

/**
 * Persistent storage for Sentry-related user preferences. Snapshot-at-boot:
 * each process reads on its own cold start, so toggle changes only take effect
 * after the next launch. Flipping a toggle to `false` also wipes the on-disk
 * Sentry envelope cache so events queued in the current session never ship.
 *
 * Persistence goes through a [Store] so unit tests can back it with a plain
 * map instead of a real `SharedPreferences` (unmocked on the JVM unit-test
 * classpath). [open] is the production constructor.
 */
internal class ComapeoPrefs(
    private val store: Store,
    private val defaults: Defaults,
    /** Wall clock; injectable so the debug-auto-off tests don't depend on real time. */
    private val now: () -> Long = { System.currentTimeMillis() },
) {
    /**
     * Minimal persistence surface. `null` from a getter means "key absent"
     * (so the caller falls back to a default). Production wraps
     * `SharedPreferences`; tests use an in-memory map.
     */
    internal interface Store {
        fun getBoolean(key: String): Boolean?
        fun putBoolean(key: String, value: Boolean)
        fun getLong(key: String): Long?
        fun putLong(key: String, value: Long)
        fun remove(key: String)
    }

    data class Defaults(
        val diagnosticsEnabled: Boolean,
        val applicationUsageData: Boolean,
        val debug: Boolean,
    )

    fun readDiagnosticsEnabled(): Boolean =
        store.getBoolean(KEY_DIAGNOSTICS_ENABLED) ?: defaults.diagnosticsEnabled

    fun writeDiagnosticsEnabled(value: Boolean) {
        store.putBoolean(KEY_DIAGNOSTICS_ENABLED, value)
    }

    fun readApplicationUsageData(): Boolean =
        store.getBoolean(KEY_APPLICATION_USAGE_DATA) ?: defaults.applicationUsageData

    fun writeApplicationUsageData(value: Boolean) {
        store.putBoolean(KEY_APPLICATION_USAGE_DATA, value)
    }

    /**
     * Read the `debug` toggle, applying the [DEBUG_MAX_AGE_MS] auto-off: if
     * debug was switched on longer ago than that, flip it off, clear the
     * timestamp, queue a `comapeo.debug.auto_disabled` breadcrumb, and return
     * `false`. A `debug=true` cell with no timestamp (e.g. enabled via the
     * configured default) is treated as "enabled now" and stamped on first read.
     *
     * The window is wall-clock based (it must survive process restarts), so a
     * backward clock change is treated conservatively: an enable timestamp in
     * the future expires debug rather than extending it. This is a best-effort
     * privacy window on the user's own device, not a security boundary.
     */
    fun readDebugEnabled(): Boolean {
        val stored = store.getBoolean(KEY_DEBUG) ?: defaults.debug
        if (!stored) return false
        val enabledAt = store.getLong(KEY_DEBUG_ENABLED_AT_MS)
        if (enabledAt == null) {
            // No recorded start (e.g. enabled via the default): start the clock.
            store.putLong(KEY_DEBUG_ENABLED_AT_MS, now())
            return true
        }
        val age = now() - enabledAt
        if (age < 0 || age > DEBUG_MAX_AGE_MS) {
            store.putBoolean(KEY_DEBUG, false)
            store.remove(KEY_DEBUG_ENABLED_AT_MS)
            DebugAutoOff.queueBreadcrumb()
            return false
        }
        return true
    }

    /**
     * Write `debug`, stamping (true) or clearing (false) the enable timestamp
     * synchronously so the window starts/stops with the value. Re-writing
     * `true` refreshes the window.
     */
    fun writeDebugEnabled(value: Boolean) {
        store.putBoolean(KEY_DEBUG, value)
        if (value) {
            store.putLong(KEY_DEBUG_ENABLED_AT_MS, now())
        } else {
            store.remove(KEY_DEBUG_ENABLED_AT_MS)
        }
    }

    companion object {
        const val PREFS_NAME = "com.comapeo.core.prefs"
        const val KEY_DIAGNOSTICS_ENABLED = "sentry.diagnosticsEnabled"
        const val KEY_APPLICATION_USAGE_DATA = "sentry.applicationUsageData"
        const val KEY_DEBUG = "sentry.debug"
        const val KEY_DEBUG_ENABLED_AT_MS = "sentry.debugEnabledAtMs"

        const val DEFAULT_DIAGNOSTICS_ENABLED = true
        const val DEFAULT_APPLICATION_USAGE_DATA = false
        const val DEFAULT_DEBUG = false

        /** 72h in milliseconds — debug mode auto-disables this long after enable. */
        const val DEBUG_MAX_AGE_MS = 72L * 60 * 60 * 1000

        @JvmStatic
        fun open(context: Context): ComapeoPrefs {
            val sentryConfig = SentryConfig.loadFromManifest(context)
            val defaults = Defaults(
                diagnosticsEnabled = sentryConfig?.diagnosticsEnabledDefault
                    ?: DEFAULT_DIAGNOSTICS_ENABLED,
                applicationUsageData = sentryConfig?.applicationUsageDataDefault
                    ?: DEFAULT_APPLICATION_USAGE_DATA,
                debug = sentryConfig?.debugDefault ?: DEFAULT_DEBUG,
            )
            val sp = context.getSharedPreferences(PREFS_NAME, Context.MODE_PRIVATE)
            return ComapeoPrefs(SharedPrefsStore(sp), defaults)
        }

        /**
         * Delete sentry-android's on-disk cache (`<cacheDir>/sentry/` — the SDK's
         * documented default) so a `diagnosticsEnabled=false` flip can't ship
         * anything from the current session on next launch. Best-effort: a
         * filesystem error never blocks the privacy opt-out.
         */
        @JvmStatic
        fun wipeSentryOutbox(context: Context) {
            wipeSentryOutboxAt(File(context.cacheDir, "sentry"))
        }

        /** Path-taking variant for unit testing — production uses the [Context] overload. */
        @JvmStatic
        fun wipeSentryOutboxAt(dir: File) {
            try {
                dir.deleteRecursively()
            } catch (_: Throwable) {
                // Swallowed — privacy opt-out is best-effort.
            }
        }
    }

    /**
     * `commit = true` (not `apply`) so a subsequent [wipeSentryOutbox] is
     * guaranteed to see a durable value on disk. Callers run from
     * AsyncFunction coroutines so the sync cost is acceptable.
     */
    private class SharedPrefsStore(private val sp: SharedPreferences) : Store {
        override fun getBoolean(key: String): Boolean? =
            if (sp.contains(key)) sp.getBoolean(key, false) else null

        override fun putBoolean(key: String, value: Boolean) {
            sp.edit(commit = true) { putBoolean(key, value) }
        }

        override fun getLong(key: String): Long? =
            if (sp.contains(key)) sp.getLong(key, 0L) else null

        override fun putLong(key: String, value: Long) {
            sp.edit(commit = true) { putLong(key, value) }
        }

        override fun remove(key: String) {
            sp.edit(commit = true) { remove(key) }
        }
    }
}

/**
 * Holds the `comapeo.debug.auto_disabled` breadcrumb queued by the debug
 * auto-off. [ComapeoPrefs.readDebugEnabled] runs before `Sentry.init`, so the
 * breadcrumb can't be added directly; it's drained by [SentryFgsBridge.init] /
 * RN init once the SDK is up.
 *
 * Known gap: the main app process and the `:ComapeoCore` FGS process are
 * separate JVMs with separate `DebugAutoOff` statics, and only the FGS side
 * drains it. If the main-process read ([ComapeoCoreModule]) performs the
 * auto-off first, the FGS read then sees `debug` already false and queues
 * nothing, so that one breadcrumb is dropped. Consequence is cosmetic — the
 * auto-off itself still happens; only the "when it turned off" timeline marker
 * is missing for that launch. Delivering it would need cross-process plumbing
 * (expose the pending flag to JS and drain it in the RN `initSentry` path).
 */
internal object DebugAutoOff {
    @Volatile
    var pending: Boolean = false
        private set

    fun queueBreadcrumb() {
        pending = true
    }

    /** Consume the pending flag; returns whether a breadcrumb is owed. */
    fun consume(): Boolean {
        val owed = pending
        pending = false
        return owed
    }
}
